### PR TITLE
Standarizes threads

### PR DESCRIPTION
**Pull request by Aranclanos**
### Purpose
Adds the SicknessThread, ElectricalThread and RadiationThread classes, all childs of ThreadBehaviour.
The atmos thread will now be a child of ThreadBehaviour.
Cleans up the managers of all these threads.

edit:
Adds GUI buttons for the ThreadBehaviour to start, stop and change the thread of a thread process to manual, mainthread and threaded.

### Notes:
~~Draft because I disabled the AtmosManagerEditor and ElectricalManagerEditor, their features should be included into the ThreadBehaviour. (stop, start, change speed, set thread to mainthread instead)~~
edit: readded on last commit

I also removed some unused code in OverlapDestroy.cs, most of that funcion was commented out
It also requires a bit more testing than what it received for the moment

The plan behind this is to include checkpoints for the infintelooptracker since we're assuming that the server crashes are related to threaded locks never stopping, due to a mistake or an infinite loop inside the thread

### CODE REVIEW

**Comment by Aranclanos** [comment]
new commit: Adds GUI buttons for the ThreadBehaviour to start, stop and change the thread of a thread process to manual, mainthread and threaded.

its finished!

File: UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs [hunk 1]
```diff
@@ -1,82 +1,162 @@
...
 {
-	[Header("Threaded Manager")] public bool IsRunning;
+	public bool running { get; private set; }
+	public uint ticker;
+
+	[Range(1, 5000)]
+	public int tickSpeed = 100;
+
+	private Thread workingThread;
+	public static List<ThreadedBehaviour> currentThreads = new List<ThreadedBehaviour>();
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs]:
```suggestion
	public static readonly List<ThreadedBehaviour> currentThreads = new List<ThreadedBehaviour>();
```

**Comment by corp-0** [comment]
Can you give the new types a namespace? They need a home.

File: UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs [hunk 2]
```diff
@@ -1,82 +1,162 @@
...
+	{
+		if (workingThread != null)
 		{
-			WorkingThread.Abort();
-			WorkingThread = null;
+			workingThread.Abort();
+			workingThread = null;
 		}
+		currentThreads.Remove(this);
+		Logger.LogFormat("<b>{0}</b> Stopped", Category.Threading, GetType().Name); ;
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs]:
```suggestion
		Logger.LogFormat("<b>{0}</b> Stopped", Category.Threading, GetType().Name);
```

File: UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+using System.Diagnostics;
+using UnityEngine;
+using UnityEngine.Profiling;
+using Managers;
+
+namespace Systems.Radiation
+{
+	public class RadiationThread : ThreadedBehaviour
+	{
+		private HashSet<Vector2Int> CircleCircumference = new HashSet<Vector2Int>();
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs]:
```suggestion
		private readonly HashSet<Vector2Int> CircleCircumference = new HashSet<Vector2Int>();
```

File: UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		private HashSet<RadiationNode> CircleArea = new HashSet<RadiationNode>();
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs]:
```suggestion
		private readonly HashSet<RadiationNode> CircleArea = new HashSet<RadiationNode>();
```

File: UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+		private List<RadiationPulse> WorkingPulseQueue = new List<RadiationPulse>();
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs]:
```suggestion
		private readonly List<RadiationPulse> WorkingPulseQueue = new List<RadiationPulse>();
```

File: UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+		private Stopwatch StopWatchlog = new Stopwatch();
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs]:
```suggestion
		private readonly Stopwatch StopWatchlog = new Stopwatch();
```

File: UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+				if (d > 0)
+				{
+					y--;
+					d = d + 4 * (x - y) + 10;
+				}
+				else
+					d = d + 4 * x + 6;
+
+				drawCircle(xc, yc, x, y);
+				//delay(50);
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Systems/Radiation/RadiationThread.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs [hunk 8]
```diff
@@ -1,82 +1,162 @@
...
+	/// The thread loop, it will constantly sleep and call the main tick
+	/// </summary>
+	private void ThreadLoop()
+	{
+		running = true;
+		Profiler.BeginThreadProfiling("Unitystation", threadName);
+		while (running && threadMode == ThreadMode.Threaded && midTick == false)
+		{
+			RunTick();
+			Thread.Sleep(tickSpeed);
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs]: We probably want the inverse of `tickSpeed` here to make sense as a speed. More like a tick delay currently.

**Review by NoooneyDude — DISMISSED** [review]
Namespaces would be great, and there is a distinct lack of PascalCase ; )

File: UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs (hunk 8 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs]: Need to take account for time Spent running RunTick

Comment (corp-0) [reply UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs]: what's the state of thsi conversation?

Comment (Aranclanos) [reply UnityProject/Assets/Scripts/Core/Threading/ThreadBehaviour.cs]: changed to tickdelay

**Comment by Aranclanos** [comment]
updated, apologies for taking so long!

**Review by Bod9001 — APPROVED** [review]

**Comment by corp-0** [comment]
I know you weren't the author of the naming issues but god it would be nice if you fixed them

**Review by NoooneyDude — APPROVED** [review]

**Pull request merged**